Add a module that loads GA credentials on a rotation

jmhooper opened the pull request:
This PR is a swing at #183. The code in here is rough and untested, but demonstrates how multiple GA credentials can be stored as environment variables.

I'm opening the PR early for input about whether we think this is the right direction?

##### Commit Message:

This commit adds a function that provides secret keys and report emails on rotation. This is necessary because we have to maintain several sets of DAP credentials to stay under Google's API rate limits.

This commit lets the analytics reporter call a function to get a set of creds which are provided on rotation. This means an even dispersement of creds amongst API calls.

The creds are loaded by looking for environment variables with the `ANALYTICS_KEY` and `ANALYTICS_REPORT_EMAIL` prefix and then grouping those based on suffix. So, for example, `ANALYTICS_KEY_1` and `ANALYTICS_REPORT_EMAIL_1` would be grouped together. Once the creds are grouped, they are stored in an array. When this new function is called, it returns the element that succeeds the last element it returned. At the end of the array, it goes back to the beginning.

konklone commented:
I think the underlying approach of rotating credentials used in a deterministic cycled way is good. The naming convention for env vars is not bad, though I wonder if there's an alternative that doesn't rely on regexing env vars, which just feels risky in an abstract way.

How would a single env var with JSON array syntax, or just comma separation (CSV!) sound? 

This also has the effect of being more easily able to detect a configuration error. If someone messes up the syntax of a JSON array or comma-separated list, this can be flagged at run-time with a descriptive error message. 

But if someone messes up the env var pattern of one of the credentials, it'll fail silently and just not be detected as a credential (unless you literally scanned all env vars looking for likely mistakes, which sounds infeasible).

So, I would recommend something that looks a bit gnarlier than the `_1`, `_2` approach, for a bit more resilience to human error.

jmhooper commented:
Sweet. I had the same thought. My concerns about using JSON were that:

1. It's trickier to escape everything properly
2. As the number of creds grow, the output when you inspect the env is uglier

I don't bring up either of those as reasons not to use JSON, but rather as things we should think about. I think JSON has the advantage of being more straight forward than an env var pattern. I think of the options it's easier to explain how to configure in the README which imo is a good test for which path is better.

I'll pivot to using a single ENV var which provides credentials in a JSON array.

konklone commented:
:+1: Also, comma-separated might make sense if there are guaranteed to be no commas in the creds.

jmhooper commented:
I've changed over to a JSON encoded array. I've also added tests and integrated with the rest of the reporter.

This is no longer WIP and is ready for review.

konklone commented:
This looks good to me. Merging.

Two things that come to mind for the near future:

* It might be nice, as a convenience, to have a CLI function to generate JSON to copy/paste into the env var, so that people don't have to edit JSON by hand.
* I believe this work creates breaking changes for those using `analytics-reporter` in a PaaS context. The City of Philadelphia is the one example I know for sure, but it's conceivable that others are using an env var based approach for deployment. This might be a good time to get a handle on PoCs for downstream users, and perhaps create expectations around contributions and how the maintainers can reliably contact production users to discuss/notify changes. cc @laurenancona for perspective.

laurenancona commented:
@konklone Just catching up on GH notifs - I _think_ I mentioned something like this to @gbinal & @jmhooper during a DAP weekly coworking session last week. I would definitely like to help ensure we both communicate breaking changes to current users (and possible mitigations), and discuss potential route for such notifications in the future - acknowledging this might be a larger philosophical conversation. 

Keeping to a tighter SemVer release schema + explicit notes is the first thing that comes to mind - thoughts?

jmhooper commented:
> I believe this work creates breaking changes for those using analytics-reporter in a PaaS context

I think that was the case when we were using the env var prefixes, but the `DAP_CREDENTIALS` env var is not something we've used in the past to my knowledge. With that said, it does introduce new functionality so I should have bumped the minor version in this PR.
